### PR TITLE
Add `SqliteStore` backend

### PR DESCRIPTION
**Pull request by tnull**
~~Based on #101.~~

Finally, we add an SQLite-based `KVStore` implementation which becomes the new default
default for `Node::build`. The `FilesystemStore` is still configurable via `Node::build_with_fs_store`

### CODE REVIEW

**Comment by G8XSU** [comment]
Can we move the last commit of SQLite impl is separate PR, would like to review it separately. :)

File: src/io/sqlite_store.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+/// [SQLite]: https://sqlite.org
+pub struct SqliteStore {
+	connection: Arc<Mutex<Connection>>,
+}
+
+impl SqliteStore {
+	pub(crate) fn new(dest_dir: PathBuf) -> Self {
+		let msg =
+			format!("Failed to create database destination directory: {}", dest_dir.display());
+		fs::create_dir_all(dest_dir.clone()).expect(&msg);
```
Comment (jkczyz) [inline src/io/sqlite_store.rs:34]: Consider using a variant where you don't need to allocate the string for the successful case. Likewise throughout.

File: src/io/sqlite_store.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+/// A [`KVStore`] implementation that writes to and reads from an [SQLite] database.
+///
+/// [SQLite]: https://sqlite.org
+pub struct SqliteStore {
+	connection: Arc<Mutex<Connection>>,
+}
+
+impl SqliteStore {
+	pub(crate) fn new(dest_dir: PathBuf) -> Self {
```
Comment (jkczyz) [inline src/io/sqlite_store.rs:31]: Should this return an error so that any failure doesn't cause a mobile app to crash? I'm assuming that's what `expect` will do, though I'm not sure if there is any way to catch this in other languages.

File: src/io/sqlite_store.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+		))?;
+		let namespace = parent_directory.display().to_string();
+
+		let dest_without_namespace = dest_file_path
+			.strip_prefix(&namespace)
+			.map_err(|_| lightning::io::Error::new(lightning::io::ErrorKind::InvalidInput, msg))?;
+		let key = dest_without_namespace.display().to_string();
+
+		self.write(&namespace, &key, &object.encode())?;
+		Ok(())
```
Comment (jkczyz) [inline src/io/sqlite_store.rs]: Can make this one expression?

File: src/io/sqlite_store.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+			let sqlite_store = SqliteStore::new(rand_dir.into());
+			let namespace = "testspace";
+			let key = "testkey";
+
+			// Test the basic KVStore operations.
+			sqlite_store.write(namespace, key, &data).unwrap();
+
+			let listed_keys = sqlite_store.list(namespace).unwrap();
+			assert_eq!(listed_keys.len(), 1);
+			assert_eq!(listed_keys[0], "testkey");
```
Comment (jkczyz) [inline src/io/sqlite_store.rs:219]: Re-use `key`.

File: src/io/sqlite_store.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::test::utils::random_storage_path;
+	use lightning::util::persist::KVStorePersister;
+	use lightning::util::ser::Readable;
+
+	use proptest::prelude::*;
+	proptest! {
```
Comment (jkczyz) [inline src/io/sqlite_store.rs:205]: Perhaps I'm misunderstanding, but what is the value of using `proptest` here. The data written doesn't seem relevant to exercising the code paths.

File: src/lib.rs [hunk 6]
```diff
@@ -320,11 +321,21 @@ impl Builder {
 		self
 	}
 
+	/// Builds a [`Node`] instance with a [`SqliteStore`] backend and according to the options
+	/// previously configured.
+	pub fn build(&self) -> Arc<Node<SqliteStore>> {
+		let storage_dir = self.config.storage_dir_path.clone();
+		fs::create_dir_all(storage_dir.clone()).expect("Failed to create LDK data directory");
+		let kv_store = Arc::new(SqliteStore::new(storage_dir.into()).into());
+		self.build_with_store(kv_store)
```
Comment (jkczyz) [inline src/lib.rs:330]: Does `build_with_store` need to be `pub`?

File: src/lib.rs [hunk 7]
```diff
@@ -473,8 +479,12 @@ impl Builder {
...
 			Ok(monitors) => monitors,
 			Err(e) => {
-				log_error!(logger, "Failed to read channel monitors: {}", e.to_string());
-				panic!("Failed to read channel monitors: {}", e.to_string());
+				if e.kind() == std::io::ErrorKind::NotFound {
+					Vec::new()
+				} else {
+					log_error!(logger, "Failed to read channel monitors: {}", e.to_string());
+					panic!("Failed to read channel monitors: {}", e.to_string());
+				}
```
Comment (jkczyz) [inline src/lib.rs:487]: Looks like this should be a separate commit?

File: src/lib.rs [hunk 8]
```diff
@@ -320,11 +321,21 @@ impl Builder {
 		self
 	}
 
+	/// Builds a [`Node`] instance with a [`SqliteStore`] backend and according to the options
+	/// previously configured.
+	pub fn build(&self) -> Arc<Node<SqliteStore>> {
+		let storage_dir = self.config.storage_dir_path.clone();
+		fs::create_dir_all(storage_dir.clone()).expect("Failed to create LDK data directory");
```
Comment (jkczyz) [inline src/lib.rs:328]: Could you DRY this up and make a separate function to use in both methods?

File: src/test/functional_tests.rs [hunk 9]
```diff
@@ -321,6 +321,55 @@ fn start_stop_reinit() {
 	reinitialized_node.stop().unwrap();
 }
 
+#[test]
+fn start_stop_reinit_fs_store() {
```
Comment (jkczyz) [inline src/test/functional_tests.rs:325]: Should we also do this for `SqliteStore`?

File: src/io/sqlite_store.rs [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+		#[test]
+		fn read_write_remove_list_persist(data in any::<[u8; 32]>()) {
```
Comment (G8XSU) [inline src/io/sqlite_store.rs:207]: can we refactor this testcase so that all impl use same set of tests ?

File: src/lib.rs [hunk 11]
```diff
@@ -320,15 +321,29 @@ impl Builder {
...
+		let kv_store = Arc::new(SqliteStore::new(storage_dir.into()).into());
+		self.build_with_store(kv_store)
+	}
 
-		let ldk_data_dir = format!("{}/ldk", config.storage_dir_path);
-		fs::create_dir_all(ldk_data_dir.clone()).expect("Failed to create LDK data directory");
+	/// Builds a [`Node`] instance with a [`FilesystemStore`] backend and according to the options
+	/// previously configured.
+	pub fn build_with_fs_store(&self) -> Arc<Node<FilesystemStore>> {
```
Comment (G8XSU) [inline src/lib.rs:335]: can't we do similar to set_esplora_server_url ? 
and use same builder pattern for this.

Comment (G8XSU) [reply src/lib.rs:335]: moved to another PR

File: src/io/sqlite_store.rs [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+		connection
+			.pragma(Some(rusqlite::DatabaseName::Main), "user_version", SCHEMA_USER_VERSION, |_| {
+				Ok(())
+			})
+			.expect(&msg);
+
+		let sql = format!(
+			"CREATE TABLE IF NOT EXISTS {} (
+			namespace TEXT NOT NULL,
+			key TEXT NOT NULL,
```
[the remark this answers is not in the document]
Comment (G8XSU) [reply src/io/sqlite_store.rs]: i think we should add the non-empty constraint on namespace as well, it is unexpected to have empty namespace, specially when we allow for clients to use custom schema with pluggable backend kvStore. 

And instead use a static namespace in place of an empty one where needed.

File: src/io/sqlite_store.rs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+			keys.push(k.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?);
+		}
+
+		Ok(keys)
+	}
+}
+
+impl KVStorePersister for SqliteStore {
+	fn persist<W: Writeable>(&self, prefixed_key: &str, object: &W) -> lightning::io::Result<()> {
+		let msg = format!("Could not persist file for key {}.", prefixed_key);
```
[the remark this answers is not in the document]
Comment (G8XSU) [reply src/io/sqlite_store.rs]: "having the different backends decide to rename namespaces is a bad idea". yes exactly
that is why we should not be using anything to do with parent directories and key parsing from ldk-key as namespace keys. (for all kvstores including filesystem)

afaiu, ldk doesn't enforce anything by having keys like "monitors/[txid][index]", ldk is just providing a prefix.
 we could have something like "ldk_core" as namespace and "monitors/[txid][index]" as full key. 
 
 otherwise we are letting ldk-core create namespaces dynamically instead of ldk-node having fixed set of namespaces.(even though we own ldk-core, this is not ideal from ldk-node user perspective)
 
otherwise, in future, if ldk decides to store something new as "xyzMeta/" we would end up creating a namespace out of it and migration code will need to change.

File: src/io/sqlite_store.rs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+		connection
+			.pragma(Some(rusqlite::DatabaseName::Main), "user_version", SCHEMA_USER_VERSION, |_| {
+				Ok(())
+			})
+			.unwrap_or_else(|_| panic!("Failed to set PRAGMA user_version"));
+
+		let sql = format!(
+			"CREATE TABLE IF NOT EXISTS {} (
+			namespace TEXT NOT NULL,
```
Comment (G8XSU) [inline src/io/sqlite_store.rs:49]: i think we should add the non-empty constraint on namespace as well, it is unexpected to have empty namespace, specially when we allow for clients to use custom schema with pluggable backend kvStore.

And instead use a static namespace in place of an empty one where needed.

File: src/io/sqlite_store.rs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+			key TEXT NOT NULL CHECK (length(key) > 0),
```
Comment (G8XSU) [inline src/io/sqlite_store.rs:50]: not empty check should look like 
CHECK(key <> '') instead of checking length

File: src/io/sqlite_store.rs [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+					":namespace": namespace,
+					":key": key,
+					":value": buf,
+				},
+			)
+			.map_err(|e| {
+				let msg = format!("Failed to write to key {}/{}: {}", namespace, key, e);
+				std::io::Error::new(std::io::ErrorKind::Other, msg)
+			})?;
+		Ok(())
```
Comment (jkczyz) [inline src/io/sqlite_store.rs:116]: Can also remove this explicit `Ok(())`.

File: src/io/sqlite_store.rs (hunk 15 above)
Comment (G8XSU) [reply src/io/sqlite_store.rs]: its a nit but depending on database constraint optimization, length might need to compute length before comparison.

File: src/io/sqlite_store.rs (hunk 14 above)
Comment (G8XSU) [reply src/io/sqlite_store.rs:49]: why do we want to have this inconsistency/workaround instead of opting for a simpler thing ?

File: src/io/sqlite_store.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (G8XSU) [reply src/io/sqlite_store.rs]: >No, it's clearly not just a prefix but an implied namespace, actually even an implied path as all assumptions are made around FilesystemPersister

there is no enforced implied namespace, cash-lit doesn't care about it, ldk doesn't care about it. We are making that assumption right now in this PR, as we split the key into a file-path for a non-filesystem-directory datastore. We shouldn't be thinking of this having any relation to file-path or filesystem-directory/subdirectory.

File: src/io/sqlite_store.rs [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+			})?);
+		}
+
+		Ok(keys)
+	}
+}
+
+impl KVStorePersister for SqliteStore {
+	fn persist<W: Writeable>(&self, prefixed_key: &str, object: &W) -> lightning::io::Result<()> {
+		let (namespace, key) = get_namespace_and_key_from_prefixed(prefixed_key)?;
```
Comment (G8XSU) [inline src/io/sqlite_store.rs:173]: We shouldn't split the key into a file-path for a non-filesystem-directory datastore. We shouldn't be thinking of this having any relation to file-path or filesystem-directory/subdirectory. This is a flawed assumption reminiscent of filesystem-datastore.

as being discussed in ([link](https://github.com/lightningdevkit/ldk-node/pull/100#discussion_r1207362176))

**Comment by jkczyz** [comment]
> @G8XSU @jkczyz Can I squash the fixups here?

SGTM

**Review by G8XSU — APPROVED** [review]
Overall LGTM

**Review by jkczyz — APPROVED** [review]

**Comment by imaginator** [comment]
What's the reasoning behind having KV storage instead of a full SQL schema?
